### PR TITLE
Issue #474 : Fixing Caching tiers should flush all cached entries…

### PR DESCRIPTION
**Pull request by rishabhmonga**
… on shutdown

### CODE REVIEW

File: impl/src/main/java/org/ehcache/internal/store/tiering/CacheStore.java [hunk 1]
```diff
@@ -383,13 +383,15 @@ public void disableStoreEventNotifications() {
...
+    @SuppressWarnings("unchecked")
     @Override
-    public void releaseStore(Store<?, ?> resource) {
+    public void releaseStore(Store<?, ?> resource) throws CacheAccessException {
       Map.Entry<CachingTier.Provider, AuthoritativeTier.Provider> entry = providersMap.get(resource);
       if (entry == null) {
         throw new IllegalArgumentException("Given store is not managed by this provider : " + resource);
       }
       CacheStore cacheStore = (CacheStore) resource;
+      cacheStore.flushAll((Store)cacheStore.realCachingTier);
```
Comment (AbfrmBlr) [inline impl/src/main/java/org/ehcache/internal/store/tiering/CacheStore.java]: With #487 this assumption may not be true anymore.

Comment (ljacomet) [inline impl/src/main/java/org/ehcache/internal/store/tiering/CacheStore.java]: Had a chat with Rishabh on this, I believe the right path is to make use of `CachingTier.invalidate()`

**Comment by rishabhmonga** [comment]
test this please

File: impl/src/test/java/org/ehcache/internal/store/tiering/CacheStoreTest.java [hunk 2]
```diff
@@ -80,8 +83,9 @@ public void testGetHitsAuthoritativeTier() throws Exception {
     when(cachingTier.getOrComputeIfAbsent(any(Number.class), any(Function.class))).then(new Answer<Store.ValueHolder<CharSequence>>() {
       @Override
       public Store.ValueHolder<CharSequence> answer(InvocationOnMock invocation) throws Throwable {
-        Number key = (Number) invocation.getArguments()[0];
```
Comment (ljacomet) [inline impl/src/test/java/org/ehcache/internal/store/tiering/CacheStoreTest.java]: Can you revert all these formatting changes and tweak the config of your IDE to only reformat code you touched?
They add noise to the PR for no added value. Same holds for the next 4 blocks of changes.

**Comment by rishabhmonga** [comment]
test this please

File: impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java [hunk 3]
```diff
@@ -367,6 +367,16 @@ public void clear() throws CacheAccessException {
 
   @Override
   public void invalidate() throws CacheAccessException {
+    if(map.map != null) {
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java]: The whole invalidation process may take a while. Should we improve this part so that it can happens without interference from other cache operations? I am interested in having other opinions on this.

One idea would be to swap the underlying map while we do the operation as it can be discarded anyway.

File: impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java [hunk 4]
```diff
@@ -367,6 +367,16 @@ public void clear() throws CacheAccessException {
 
   @Override
   public void invalidate() throws CacheAccessException {
```
Comment (ljacomet) [inline impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java]: With the change of contract, the call from `CacheStore.clear()` seems wrong. We may need to add a `clear()` method on the `CachingTier` interface. Others?

File: impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java (hunk 3 above)
Comment (chrisdennis) [inline impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java]: Do we ever invalidate when other cache operations are happening?

**Comment by rishabhmonga** [comment]
test this please

File: impl/src/test/java/org/ehcache/internal/store/tiering/CacheStoreTest.java (hunk 2 above)
Comment (ljacomet) [inline impl/src/test/java/org/ehcache/internal/store/tiering/CacheStoreTest.java]: Looks like this updated changeset revived the unrelated formatting changes :smile:

File: impl/src/test/java/org/ehcache/internal/store/tiering/CacheStoreTest.java [hunk 5]
```diff
@@ -270,7 +273,7 @@ public void testClear() throws Exception {
 
     cacheStore.clear();
 
-    verify(cachingTier, times(1)).invalidate();
+//    verify(cachingTier, times(1)).invalidate();
```
Comment (ljacomet) [inline impl/src/test/java/org/ehcache/internal/store/tiering/CacheStoreTest.java]: Please do not leave commented out code

**Comment by ljacomet** [comment]
Minor remarks otherwise matched what was discussed

**Comment by rishabhmonga** [comment]
test this please

**Comment by rishabhmonga** [comment]
test this please

File: impl/src/test/java/org/ehcache/internal/store/tiering/CacheStoreTest.java [hunk 6]
```diff
@@ -349,8 +351,8 @@ public void testComputeIfAbsent_computes() throws Exception {
     when(authoritativeTier.computeIfAbsentAndFault(any(Number.class), any(Function.class))).thenAnswer(new Answer<Store.ValueHolder<CharSequence>>() {
       @Override
       public Store.ValueHolder<CharSequence> answer(InvocationOnMock invocation) throws Throwable {
-        Number key = (Number) invocation.getArguments()[0];
```
Comment (ljacomet) [inline impl/src/test/java/org/ehcache/internal/store/tiering/CacheStoreTest.java]: This is again unrelated formatting changes - can you get rid of them?

**Comment by rishabhmonga** [comment]
test this please

File: core/src/main/java/org/ehcache/spi/cache/Store.java [hunk 7]
```diff
@@ -441,7 +441,7 @@
      * Informs this Provider, a Store it created is being disposed (i.e. closed)
      * @param resource
      */
-    void releaseStore(Store<?, ?> resource);
+    void releaseStore(Store<?, ?> resource) throws CacheAccessException;
```
Comment (lorban) [inline core/src/main/java/org/ehcache/spi/cache/Store.java]: I'm not sure rethrowing CacheAccessException is the right thing to do here. IMHO the releaseStore() implementations should handle those internally.

File: core/src/main/java/org/ehcache/spi/cache/tiering/CachingTier.java [hunk 8]
```diff
@@ -82,7 +82,7 @@
   interface Provider extends Service {
     <K, V> CachingTier<K, V> createCachingTier(Store.Configuration<K, V> storeConfig, ServiceConfiguration<?>... serviceConfigs);
 
-    void releaseCachingTier(CachingTier<?, ?> resource);
+    void releaseCachingTier(CachingTier<?, ?> resource) throws CacheAccessException;
```
Comment (lorban) [inline core/src/main/java/org/ehcache/spi/cache/tiering/CachingTier.java]: Same remark as above: do we want this exception to be rethrown or handled internally?

File: core/src/main/java/org/ehcache/spi/cache/Store.java (hunk 7 above)
Comment (chrisdennis) [inline core/src/main/java/org/ehcache/spi/cache/Store.java]: Anything thrown here is going to propagate all the the way out the the user wrapped in a `StateTransitionException`.  I'm not averse to having these methods throw when bad things happen, but to me two things need to happen.  Firstly we can't be throwing eagerly, and secondly I don't think we should be throwing CacheAccessException... it's an internal exception that we shouldn't be leaking (even if it is wrapped).  Does this sound sensible?

Comment (alexsnaps) [inline core/src/main/java/org/ehcache/spi/cache/Store.java]: Right, and there is logic in those state transitions, so _I think_ in this case, the `.shutdown()` will continue. If anything though, I agree with @chrisdennis and we shouldn't make throwing `CacheAccessException` part of the contract, as it's an internal, implementation specific detail that the `Store` is even accessed further when `releaseStore` is invoked.

File: impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java [hunk 9]
```diff
@@ -366,8 +366,17 @@ public void clear() throws CacheAccessException {
...
+    if(map.map != null) {
+      for (K k : map.map.keySet()) {
+        invalidate(k);
+      }
+    }
+    if(map.keyCopyMap != null) {
+      for(OnHeapKey<K> key : map.keyCopyMap.keySet()) {
+        invalidate(key.getActualKeyObject());
+      }
+    }
```
Comment (alexsnaps) [inline impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java]: This can't possibly happen! See the discussion on @lorban PR, this method _should_ be a plain clear!

**Comment by rishabhmonga** [comment]
test this please
